apps/microtvm/arduino/template_project/microtvm_api_server.py

```diff
@@ -0,0 +1,423 @@
+
+                with filename.open("w") as file:
+                    file.writelines(lines)
+
+    def generate_project(self, model_library_format_path, standalone_crt_dir, project_dir, options):
+        # Reference key directories with pathlib
+        project_dir = pathlib.Path(project_dir)
+        project_dir.mkdir()
+        source_dir = project_dir / "src"
+        source_dir.mkdir()
+
+        # Copies files from the template folder to project_dir. model.h is copied here,
+        # but will also need to be templated later.
+        if IS_TEMPLATE:
+            shutil.copy2(API_SERVER_DIR / "microtvm_api_server.py", project_dir)
+            self._copy_project_files(API_SERVER_DIR, project_dir, options["project_type"])
+
+        # Copy standalone_crt into src folder
+        self._copy_standalone_crt(source_dir, standalone_crt_dir)
+        self._remove_unused_components(source_dir, options["project_type"])
+
+        # Unpack the MLF and copy the relevant files
+        metadata = self._disassemble_mlf(model_library_format_path, source_dir)
+        shutil.copy2(model_library_format_path, source_dir / "model")
+
+        # For AOT, template model.h with metadata to minimize space usage
+        if options["project_type"] == "example_project":
+            self._template_model_header(source_dir, metadata)
+
+        self._change_cpp_file_extensions(source_dir)
+
+        # Recursively change imports
+        self._convert_imports(project_dir, source_dir)
+
+    def _get_fqbn(self, options):
+        o = BOARD_PROPERTIES[options["arduino_board"]]
+        return f"{o['package']}:{o['architecture']}:{o['board']}"
+
+    def build(self, options):
+        BUILD_DIR.mkdir()
+        print(BUILD_DIR)
+
+        compile_cmd = [
+            options["arduino_cli_cmd"],
+            "compile",
+            "./project/",
+            "--fqbn",
+            self._get_fqbn(options),
+            "--build-path",
+            BUILD_DIR.resolve(),
+        ]
+
+        if options.get("verbose"):
+            compile_cmd.append("--verbose")
+
+        # Specify project to compile
+        output = subprocess.check_call(compile_cmd)
+        assert output == 0
+
+    # We run the command `arduino-cli board list`, which produces
+    # outputs of the form:
+    """
+    Port         Type              Board Name FQBN                          Core
+    /dev/ttyS4   Serial Port       Unknown
+    /dev/ttyUSB0 Serial Port (USB) Spresense  SPRESENSE:spresense:spresense SPRESENSE:spresense
+    """
+
+    def _auto_detect_port(self, options):
+        list_cmd = [options["arduino_cli_cmd"], "board", "list"]
+        list_cmd_output = subprocess.check_output(list_cmd).decode("utf-8")
+        # Remove header and new lines at bottom
+        port_options = list_cmd_output.split("\n")[1:-2]
+
+        # Select the first compatible board
+        fqbn = self._get_fqbn(options)
+        for port_option in port_options:
+            if fqbn in port_option:
```

Comment:
(if this is too complicated, add a TODO instead)

can you determine the exact part of each line to look at by doing 
```
fqbn_index = list_cmd_output.split("\n", 1)[0].index("FQBN")
core_index = list_cmd_output.split("\n", 1)[0].index("Core")

# ...
port_fqbn = port_option[fqbn_index:core_index].rstrip()
if fqbn == port_fqbn:
  return port_option.split(" ")[0]
```

Reply:
Yea, that's probably cleaner. Fixed!